pcgc/pcgc.go

```diff
@@ -1,20 +1,153 @@
+// Package pcgc is a service client for Cloud Manager API.
+//
+// To be able to interact with this API, you have to
+// create a new service:
+//
+//     s := pcgc.NewService(nil)
+//
+// The Service struct has all the methods you need
+// to interact with Cloud Manager API.
+//
 package pcgc
 
 import (
+	"bytes"
 	"context"
-	"fmt"
+	"encoding/json"
+	"github.com/bobziuchkovski/digest"
+	"github.com/google/go-querystring/query"
+	"io"
 	"net/http"
+	"reflect"
+	"runtime"
 )
 
-// DoGetRequest creates a new GET HTTP request, executes it, and returns the response
-func DoGetRequest(url string) (*http.Response, error) {
-	ctx, onCancel := context.WithTimeout(context.Background(), RequestTimeout)
-	defer onCancel()
+const (
+	// Version Library version
+	Version = "v1"
+	// DefaultUserAgent User agent
+	DefaultUserAgent = "pcgc/" + Version + " (" + runtime.GOOS + "; " + runtime.GOARCH + ")"
+	// DefaultURL api base url
+	DefaultURL = "http://localhost:8080"
+	// DefaultMediaType default media type
+	DefaultMediaType = "application/json"
+)
+
+// Service represents your API.
+type Service struct {
+	client   *http.Client
+	URL      string
+	user     string
+	password string
+}
+
+// NewService creates a Service using the given client, if none is provided
+// it uses http.DefaultClient.
+func NewService(c *http.Client, u string, p string) *Service {
+	if c == nil {
+		c = http.DefaultClient
+	}
+	return &Service{
+		client:   c,
+		URL:      DefaultURL,
+		user:     u,
+		password: p,
+	}
+}
+
+// NewRequest generates an HTTP request, but does not perform the request.
+func (s *Service) NewRequest(ctx context.Context, method, path string, body interface{}, q interface{}) (*http.Request, error) {
+
+	var rbody io.Reader
+	switch t := body.(type) {
+	case nil:
+	case string:
+		rbody = bytes.NewBufferString(t)
+	case io.Reader:
+		rbody = t
+	default:
+		v := reflect.ValueOf(body)
```

Comment:
Using reflection has a bunch of disbenefits.  In general, I'd like to avoid using reflection unless I absolutely have to.  For golang, that usually means duplicating code rather than trying to write something generic and ending up with something hard to read/comprehend.

Reply:
See the other PR: https://github.com/mongodb-labs/pcgc/pull/2/files#diff-98f7fa0d2792bea30ecebcad8226b0b7R78 (granted it doesn't handle digest auth yet, but it doesn't look like that's related to this code)